risc0-adapter/src/host.rs

```diff
@@ -0,0 +1,64 @@
+use std::cell::RefCell;
+
+use risc0_zkp::{core::config::HashSuiteSha256, field::baby_bear::BabyBear};
+use risc0_zkvm::{receipt::verify_with_hal, serde::to_vec, sha::Impl, Prover, Receipt};
+use serde::{Deserialize, Serialize};
+use sovereign_sdk::zk::traits::{Zkvm, ZkvmHost};
+
+use crate::Risc0MethodId;
+
+const CIRCUIT: risc0_circuit_rv32im::CircuitImpl = risc0_circuit_rv32im::CircuitImpl::new();
+
+pub struct Risc0Host<'a> {
+    prover: RefCell<Prover<'a>>,
+}
+
+impl<'a> Risc0Host<'a> {
+    pub fn new(elf: &'a [u8]) -> Self {
+        Self {
+            prover: RefCell::new(
+                Prover::new(elf).expect("Prover should be constructed from valid ELF binary"),
+            ),
+        }
+    }
+
+    pub fn run(&mut self) -> anyhow::Result<Receipt> {
+        self.prover.borrow_mut().run()
+    }
+}
+
+impl<'a> ZkvmHost for Risc0Host<'a> {
+    fn write_to_guest<T: Serialize>(&self, item: T) {
+        let serialized = to_vec(&item).expect("Serialization to vec is infallible");
+        self.prover.borrow_mut().add_input_u32_slice(&serialized);
+    }
+}
+
+impl<'prover> Zkvm for Risc0Host<'prover> {
+    type CodeCommitment = Risc0MethodId;
+
+    type Error = anyhow::Error;
+
+    fn verify<'a>(
+        serialized_proof: &'a [u8],
+        code_commitment: &Self::CodeCommitment,
+    ) -> Result<&'a [u8], Self::Error> {
+        // let journal:  = bincode::deserialize(serialized_proof)?;
+        let receipt: Risc0Proof<'a> = bincode::deserialize(serialized_proof)?;
+        verify_with_hal(
+            &risc0_zkp::verify::CpuVerifyHal::<BabyBear, HashSuiteSha256<BabyBear, Impl>, _>::new(
+                &CIRCUIT,
+            ),
+            &code_commitment.0,
+            &receipt.seal,
+            receipt.journal,
+        )?;
+        Ok(receipt.journal)
+    }
+}
+
+#[derive(Serialize, Deserialize)]
```

Comment:
According to #285 we prefere fully qualified serde derive, so you don't need to feature gate import.

Reply:
This package can't feature gate the import, so we don't need to qualify the paths to make it easier. I've qualified it just to make it uniform with the other packages though